src/synapse/workload.py

```diff
@@ -0,0 +1,467 @@
+
+    Raises:
+        CreateMjolnirConfigError: something went wrong creating mjolnir config.
+    """
+    try:
+        config = _get_mjolnir_config(access_token, room_id)
+        container.push(MJOLNIR_CONFIG_PATH, yaml.safe_dump(config), make_dirs=True)
+    except ops.pebble.PathError as exc:
+        raise CreateMjolnirConfigError(str(exc)) from exc
+
+
+def _create_pysaml2_config(charm_state: CharmState) -> typing.Dict:
+    """Create config as expected by pysaml2.
+
+    Args:
+        charm_state: Instance of CharmState.
+
+    Returns:
+        Pysaml2 configuration.
+
+    Raises:
+        EnableSAMLError: if SAML configuration is not found.
+    """
+    if charm_state.saml_config is None:
+        raise EnableSAMLError(
+            "SAML Configuration not found. "
+            "Please verify the integration between SAML Integrator and Synapse."
+        )
+
+    saml_config = charm_state.saml_config
+    entity_id = (
+        charm_state.public_baseurl
+        if charm_state.public_baseurl is not None
+        else f"https://{charm_state.server_name}"
+    )
+    sp_config = {
+        "metadata": {
+            "remote": [
+                {
+                    "url": saml_config["metadata_url"],
+                },
+            ],
+        },
+        "allow_unknown_attributes": True,
+        "service": {
+            "sp": {
+                "entityId": entity_id,
+                "allow_unsolicited": True,
+            },
+        },
+    }
+    # login.staging.ubuntu.com and login.ubuntu.com
+    # dont send uid in SAMLResponse so this will map
+    # as expected
+    if "ubuntu.com" in saml_config["metadata_url"]:
+        sp_config["attribute_map_dir"] = "/usr/local/attributemaps"
+
+    return sp_config
+
+
+def enable_saml(container: ops.Container, charm_state: CharmState) -> None:
+    """Change the Synapse configuration to enable SAML.
+
+    Args:
+        container: Container of the charm.
+        charm_state: Instance of CharmState.
+
+    Raises:
+        EnableSAMLError: something went wrong enabling SAML.
+    """
+    try:
+        config = container.pull(SYNAPSE_CONFIG_PATH).read()
+        current_yaml = yaml.safe_load(config)
+        if charm_state.public_baseurl is not None:
+            current_yaml["public_baseurl"] = charm_state.public_baseurl
+        # enable x_forwarded to pass expected headers
+        current_listeners = current_yaml["listeners"]
+        updated_listeners = [
+            {
+                **item,
+                "x_forwarded": True
+                if "x_forwarded" in item and not item["x_forwarded"]
+                else item.get("x_forwarded", False),
+            }
+            for item in current_listeners
+        ]
+        current_yaml["listeners"] = updated_listeners
+        current_yaml["saml2_enabled"] = True
+        current_yaml["saml2_config"] = {}
+        current_yaml["saml2_config"]["sp_config"] = _create_pysaml2_config(charm_state)
```

Comment:
It is somewhat confusing the read `current_yaml` being modified - at that point it isn't really current anymore, right? Perhaps consider renaming it to something else?

Reply:
Related: #52